cumulus/pallets/parachain-system/src/descendant_validation.rs

```diff
@@ -0,0 +1,628 @@
+/// - The state root of the provided relay parent does not match the expected value.
+/// - A relay header does not contain a BABE pre-digest.
+/// - A header is found with an invalid seal signature, or the authorities required to verify the
+///   signature are missing (current or next epoch).
+pub(crate) fn verify_relay_parent_descendants<H: Header>(
+	relay_state_proof: &RelayChainStateProof,
+	relay_parent_descendants: Vec<H>,
+	relay_parent_state_root: H::Hash,
+	expected_number_of_parents: u32,
+) -> Result<(), RelayParentVerificationError<H>> {
+	if relay_parent_descendants.len() != (expected_number_of_parents + 1) as usize {
+		return Err(InvalidNumberOfDescendants {
+			expected: expected_number_of_parents + 1,
+			received: relay_parent_descendants.len(),
+		});
+	}
+
+	let Ok(mut current_authorities) = relay_state_proof.read_authorities() else {
+		return Err(RelayParentVerificationError::MissingAuthorities)
+	};
+	let mut maybe_next_authorities = relay_state_proof.read_next_authorities().ok().flatten();
+
+	let mut next_expected_hash = None;
+
+	let mut required_next_authorities = false;
+
+	// Verify that the state root of the first block is the same as the one
+	// from the relay parent. In the PVF, we don't have the relay parent header hash
+	// available, so we need to use the storage root here to establish a chain.
+	if let Some(relay_parent) = relay_parent_descendants.get(0) {
+		if *relay_parent.state_root() != relay_parent_state_root {
+			return Err(RelayParentVerificationError::InvalidStateRoot {
+				expected: relay_parent_state_root,
+				found: *relay_parent.state_root(),
+			});
+		}
+	};
+
+	for (counter, mut current_header) in relay_parent_descendants.into_iter().enumerate() {
+		// Hash calculated while seal is intact
+		let sealed_header_hash = current_header.hash();
+		let relay_number = *current_header.number();
+
+		// Verify that the blocks actually form a chain
+		if let Some(expected_hash) = next_expected_hash {
+			if *current_header.parent_hash() != expected_hash {
+				return Err(RelayParentVerificationError::InvalidChainSequence {
+					expected: expected_hash,
+					found: *current_header.parent_hash(),
+					number: relay_number,
+				});
+			}
+		}
+		next_expected_hash = Some(sealed_header_hash);
+
+		log::debug!(target: crate::LOG_TARGET, "Validating header #{relay_number:?} ({sealed_header_hash:?})");
+		let Ok((authority_index, next_epoch_descriptor)) =
+			find_authority_idx_epoch_digest(&current_header)
+		else {
+			return Err(RelayParentVerificationError::MissingPredigest { hash: sealed_header_hash });
+		};
+
+		// Once we have seen a next epoch descriptor, we must always use the authorities of the
+		// next epoch.If the relay parent contains epoch descriptor, we shall not rotate
+		// authorities. As in that case the authorities in the state proof reflect the
+		// new authorities already.
+		if let Some(descriptor) = next_epoch_descriptor {
+			if counter != 0 {
```

Comment:
Why this check?

Reply:
The first block is the RP, if it has an epoch change in it we should still use the OG authorities and not next sessions.